cmd/kubectl-gadget/top/top.go

```diff
@@ -47,11 +59,153 @@ func addCommonTopFlags(
 	command *cobra.Command,
 	commonTopFlags *CommonTopFlags,
 	commonFlags *utils.CommonFlags,
-	defaultMaxRows int,
 	sortBySlice []string,
+	sortBySliceDefault []string,
 ) {
-	command.Flags().IntVarP(&commonTopFlags.MaxRows, "max-rows", "r", defaultMaxRows, "Maximum rows to print")
-	command.Flags().StringVarP(&commonTopFlags.SortBy, "sort", "", sortBySlice[0], fmt.Sprintf("Sort column (%s)", strings.Join(sortBySlice, ", ")))
-
+	command.Flags().IntVarP(&commonTopFlags.MaxRows, "max-rows", "r", top.MaxRowsDefault, "Maximum rows to print")
+	command.Flags().StringVarP(&commonTopFlags.SortBy, "sort", "", strings.Join(sortBySliceDefault, ","), fmt.Sprintf("Sort by column. Join multiple columsn with ','. Prefix with '-' to sort descending for that column. Columns: (%s)", strings.Join(sortBySlice, ", ")))
 	utils.AddCommonFlags(command, commonFlags)
 }
+
+// TopParser defines the interface that every top-gadget parser has to
+// implement.
+type TopParser[Stats any] interface {
+	// BuildColumnsHeader returns a header to be used when the user requests to
+	// present the output in columns.
+	BuildColumnsHeader() string
+	TransformStats(*Stats) string
+
+	UnmarshalStats(line, node string) ([]*Stats, error)
+	CollectStats(nodeStats map[string][]*Stats) []*Stats
+}
+
+// TopGadget represents a gadget belonging to the top category.
+type TopGadget[Stats any] struct {
+	sync.Mutex
+
+	name           string
+	commonTopFlags *CommonTopFlags
+	params         map[string]string
+	parser         TopParser[Stats]
+	nodeStats      map[string][]*Stats
+	colMap         columns.ColumnMap[Stats]
+}
+
+func (g *TopGadget[Stats]) Run(args []string) error {
+	var err error
+
+	if len(args) == 1 {
+		g.commonTopFlags.OutputInterval, err = strconv.Atoi(args[0])
+		if err != nil {
+			return commonutils.WrapInErrInvalidArg("<interval>",
+				fmt.Errorf("%q is not a valid value", args[0]))
+		}
+	} else {
+		g.commonTopFlags.OutputInterval = top.IntervalDefault
+	}
+
+	sortByColumns := strings.Split(g.commonTopFlags.SortBy, ",")
+	cols := columns.MustCreateColumns[Stats]()
```

Comment:
@flyth do you have any idea if i could use the `GetColumns` of that specific `types.go` here as well? Or how it could be solved when we have custom columns as you mentioned

Reply:
You already have a field `colMap         columns.ColumnMap[Stats]` in your `TopGadget` struct - that could be used here as well, right? As long as the initialization happens in the gadget specific code (where you can call a types.GetColumns()), it should work.

But to answer your question: I don't think it's possible to call that function generically without an interface. I would love to see an interface for all gadgets serving all kinds of metadata (like name, description, possible parameters, columns instance etc.) Hopefully we can add that later as it would help a lot with other/third party interfaces.

Reply:
Yes, i do have that in my `TopGadget`. Actually did not notice it until you mentioned it.